### PR TITLE
fix(vscode): many completion improvements

### LINKED ISSUE #1984: VSCode completion shows symbols from parent scopes that are not yet defined
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

Triggered completion list at the following position

```
let a = 1;

if true {

//^ cursor here
}

let b = 2;
```

### I expected this:

Completion list with the following content:

- `<globals>`
- a

### Instead, this happened:

Completion list with the following content:

- `<globals>`
- a
- b

### Is there a workaround?

_No response_

### Component:

IDE Extension

### Wing Version:

_No response_

### Node.js Version:

_No response_

### Platform(s):

_No response_

### Anything else?

_No response_

### LINKED ISSUE #3440: Inner struct literal completion doesn't work in inflight closures
### I tried this:

```zig
struct S { a: str; }
inflight () => {
  S {  }
   //^
};
```

### This happened:

Completions were shown as if it were outside the struct literal

### I expected this:

To show `a:`

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #3990: vscode should complete bring
### I tried this:

image

### This happened:

it didn't complete

### I expected this:

to complete to bring

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

0.26.5

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #3993: vscode shows the wrong context for cloud.ApiResponse
### I tried this:

image

### This happened:

vscode showed my the api.post signature although my cursor was on APIResonse 

### I expected this:

autocompletes the wrong stuff, it should show me the Response Properties and not the inflight

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

0.26.5

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #3994: vscode completed to cloud after `body:`
### I tried this:

image

### This happened:

as you notice the autocomplete suggest cloud after the column 

### I expected this:

don't complete

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #3995: vscode annoying completion to inflight
### I tried this:

annoying completion to inflight

image

### This happened:

it completes to inflight 

### I expected this:

not to complete

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

0.26.5

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #4000: vscode irrelevant completion on variable definition
### I tried this:

Tried to get completion for a ,variable definition:
```ts
bring cloud;

let bucket = 
        		 // ^ cursor here
```

### This happened:

image

### I expected this:

Expected to see the word `new`.
Expected to see `cloud` (already there).
Expected not to see `assert`, `log`, `throw`.

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

0.26.4

### Node.js Version

18.15.0

### Platform(s)

MacOS

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #4001: vscode no completion of type names on a blank file
### I tried this:

Wanted to call a static method `Json.stringify()`.
Started typing:
```ts
Js
// ^ cursor here
```

### This happened:

image

### I expected this:

Completion for `Json`.
The same would occur for any type (e.g. `str`, `num`, `array`, etc.).

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

0.26.4

### Node.js Version

18.15.0

### Platform(s)

MacOS

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #4013: autocomplete doesn't work when in the middle of a type
### I tried this:

[Screencast from 30-08-23 15:36:30.webm](https://github.com/winglang/wing/assets/1160578/5c48fba8-8a57-41ff-b148-534141f10a1e)

### This happened:

:point_up: 

### I expected this:

I expected `Bucket` to be suggested when trying to autocomplete from the middle of `Bucket`.

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

0.23.19

### Node.js Version

_No response_

### Platform(s)

Linux

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #4295: wrong vs code signature suggestion
### I tried this:

get signature suggestion for the `http.get` line in this code:
```ts
bring http;

class T {
  inflight do() {
    this.handle(inflight (context: str): str => {
      http.get()
      return "321";
    });
  }

  inflight handle(handler:  (str): str) {
    handler("123");
  }
}
```

### This happened:

got wrong suggestion: 
`(handler: inflight (str): str): void`
Screenshot 2023-09-26 at 21 32 25

### I expected this:

the `http.get` signature as a suggestion

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### LINKED ISSUE #4459: completion shows namespace instead of classes inside the namespace
### I tried this:

image

### This happened:

:point_up: 

### I expected this:

`JsiiClass` to be offered instead of the jsii lib name

### Is there a workaround?

_No response_

### Component

IDE Extension

### Wing Version

0.35.10

### Node.js Version

_No response_

### Platform(s)

Linux

### Anything else?

_No response_

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by MarkMcCulloh**
Was traveling recently and did random in-transit work to improve completions. Sorry for so many fixes at once, hopefully the improvement is worth it.

- Fixes #4295
signature_help does a better job of getting the inner-most function call

- Fixes #4000
Completions now filters out function that return void unless you're at the scope-level statement (e.g. it won't show up when you're expecting a plain old expression)

- Fixes #3440
The `inflight` transformer was the issue here due to its injected scope. The visitor handles this better now.

- Fixes #3993
This was really the same bug as #3440

- Fixes #3994
Partially the same bug as #3440, but also the completion thought you wanted a type because it was following a ":". Both are fixed.

- Fixes #1984
Simply made sure not to include any symbols that were defined after the current scope. Used spans for this instead of stmt indexes.

- Fixes #3990
Adds `bring` when looking for completions in the top scope

- Fixes #4013
Simplified logic to lookup types, kinda ended up fixing this by accident

- Fixes #4459
Basically the same bug as #4013

- Fixes #3995
Technically the inflight snippet is a valid completion item to provide to a struct, so I didn't remove it, but made sure to deprioritize snippets like that

- Fixes #4001
Added completions for builtin types

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingc/src/lsp/snapshots/completions/new_expression_nested.snap [hunk 1]
```diff
@@ -7,7 +7,7 @@ source: libs/wingc/src/lsp/completions.rs
...
   sortText: gg|Api
-  insertText: Api($0)
+  insertText: Api($1)
```
Comment (Chriscbr) [inline libs/wingc/src/lsp/snapshots/completions/new_expression_nested.snap:10]: What does this change mean?

File: libs/wingc/src/lsp/completions.rs [hunk 2]
```diff
@@ -490,36 +453,53 @@ fn get_current_scope_completions(
...
 	node_to_complete: &Node,
-	preceding_text: &String,
+	preceding_text: &str,
 ) -> Vec<CompletionItem> {
 	let mut completions = vec![];
 
 	// by default assume being after a colon means we're looking for a type, but this is not always true
 	let mut in_type = preceding_text.ends_with(':');
 
+	// This is either a top-
```
Comment (Chriscbr) [inline libs/wingc/src/lsp/completions.rs:463]: Looks like a half-finished comment maybe

File: libs/wingc/src/lsp/completions.rs [hunk 3]
```diff
@@ -585,22 +565,19 @@ fn get_current_scope_completions(
...
-			}
-
 			"custom_type" => in_type = true,
 			_ => {}
 		}
 	}
 
 	let found_stmt_index = scope_visitor.found_stmt_index.unwrap_or_default();
 	let found_env = types.get_scope_env(&scope_visitor.found_scope);
+	let error_buffer = if node_to_complete.has_error() { 1 } else { 0 };
```
Comment (Chriscbr) [inline libs/wingc/src/lsp/completions.rs:575]: What do 1 and 0 represent here? (they look sort of like magic numbers to me 🙂)

File: libs/wingc/src/lsp/completions.rs [hunk 4]
```diff
@@ -857,11 +931,24 @@ fn get_completions_from_namespace(
...
 		.envs
 		.iter()
 		.flat_map(|env| env.symbol_map.iter())
 		.flat_map(|(name, symbol)| format_symbol_kind_as_completion(name, &symbol.kind))
+		// ignore the builtin-functions
+		.filter(|c| {
+			if is_wing_file && c.kind == Some(CompletionItemKind::FUNCTION) {
+				c.label == UtilityFunctions::Assert.to_string()
+					&& c.label == UtilityFunctions::Log.to_string()
+					&& c.label == UtilityFunctions::Assert.to_string()
```
Comment (Chriscbr) [inline libs/wingc/src/lsp/completions.rs:947]: Can we use `UtilityFunctions::all()` to keep this code DRY?

File: libs/wingc/src/lsp/completions.rs [hunk 5]
```diff
@@ -857,11 +931,24 @@ fn get_completions_from_namespace(
...
+
+	let is_wing_file = matches!(namespace.module_path, ResolveSource::WingFile);
+
 	namespace
 		.envs
 		.iter()
 		.flat_map(|env| env.symbol_map.iter())
 		.flat_map(|(name, symbol)| format_symbol_kind_as_completion(name, &symbol.kind))
+		// ignore the builtin-functions
+		.filter(|c| {
```
Comment (Chriscbr) [inline libs/wingc/src/lsp/completions.rs:943]: I'm wondering if we should filter out all non-public symbols when searching a namespace for completions

Comment (Chriscbr) [inline libs/wingc/src/lsp/completions.rs:943]: It might be faster to filter before calling `format_symbol_kind_as_completion`.

File: libs/wingc/src/docs.rs [hunk 6]
```diff
@@ -217,7 +217,11 @@ fn render_signature_help(f: &FunctionSignature) -> String {
 		let param_type_unwrapped = param.typeref.maybe_unwrap_option();
 		let is_last = param_idx == f.parameters.len() - 1;
 
-		let param_name: &String = &param.name;
+		let param_name = if param.name.is_empty() {
+			param_idx.to_string()
+		} else {
+			param.name.clone()
+		};
```
Comment (Chriscbr) [inline libs/wingc/src/docs.rs:224]: I'm curious where this situation arises?

**Review by Chriscbr — APPROVED** [review]

File: libs/wingc/src/docs.rs (hunk 6 above)
Comment (MarkMcCulloh) [reply libs/wingc/src/docs.rs:224]: function types as annotations

e.g. hover over `x` [here](https://www.winglang.io/play/?code=bABlAHQAIAB4ADoAIAAoAG4AdQBtACkAOgAgAHYAbwBpAGQAIAA9ACAAKABhADoAIABuAHUAbQApACAAPQA%2BACAAewB9ADsA)

File: libs/wingc/src/lsp/completions.rs (hunk 5 above)
Comment (MarkMcCulloh) [reply libs/wingc/src/lsp/completions.rs:943]: Good call, that's much better

File: libs/wingc/src/lsp/completions.rs (hunk 4 above)
Comment (MarkMcCulloh) [reply libs/wingc/src/lsp/completions.rs]: Filtering to public-only fixes this as well

File: libs/wingc/src/lsp/completions.rs (hunk 5 above)
Comment (MarkMcCulloh) [reply libs/wingc/src/lsp/completions.rs]: Done, now that it's filtered based on the data straight from the env

File: libs/wingc/src/docs.rs (hunk 6 above)
Comment (Chriscbr) [reply libs/wingc/src/docs.rs:224]: Got it. Maybe name them `arg0`, `arg1`, etc. by default?

Comment (MarkMcCulloh) [reply libs/wingc/src/docs.rs]: I'm good with that. I didn't do it initially cause I was worried it would be misleading (since they don't have names) but it's probably better.

File: libs/wingc/src/lsp/completions.rs (hunk 3 above)
Comment (MarkMcCulloh) [reply libs/wingc/src/lsp/completions.rs]: Rewrote this not to need it anymore, should hopefully be more clear
